service/src/test/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiControllerConnectedTest.java

```diff
@@ -88,6 +144,368 @@ public void createBigQueryDataset_createdResourceEqualsGotResource() throws Exce
         actualBqDataset.getAttributes().getProjectId());
   }
 
+  @Test
+  public void cloneControlledBqDataset_requesterNoReadAccessOnSourceWorkspace_throws403()
```

Comment:
This set of tests looks really good. I wonder if we should partition this test file into more than one. It is going to get really big I think.

Reply:
Done. Split out into per-resource tests and updated ADDING_RESOURCES.md. You may want to take another look at ADDING_RESOURCES.md.

FYI @zloery maybe Azure could split into per-resource tests.